include/libtorrent/aux_/portmap.hpp

```diff
@@ -66,6 +66,31 @@ namespace libtorrent { namespace aux {
 	protected:
 		~portmap_callback() {}
 	};
+	struct base_mapping_t
+	{
+		enum class action : std::uint8_t { none, add, del };
+
+		// the time the port mapping will expire
+		time_point expires;
+
+		action act = action::none;
+
+		// the external (on the NAT router) port
+		// for the mapping. This is the port we
+		// should announce to others
+		int external_port = 0;
+
+		aux::portmap_protocol protocol = aux::portmap_protocol::none;
+
+		char const* protocol_name() const
```

Comment:
I think these functions really should be free functions, something like this:

``char const* to_string(aux::portmap_protocol);`` and char const* to_string(aux::action);``